src/modm/board/feather_m4/module.md

```diff
@@ -0,0 +1,76 @@
+factor. That does, however, limit the number of pins broken out on the board.
+
+Out of 51 GPIO pins on the SAMD51J19, only 22 have pads for the breakout headers.
+One of these, labeled AREF (PA03), is connected directly to 3.3V and thus cannot
+be used for IO. (This is due to a problem with the chip itself; the DAC will not
+function without this connection. If the DAC is not needed, the trace can be
+cut and the pin used as desired.)
+
+There are seven other pins not broken out but used for devices on-board:
+six are used for a 2MB QSPI Flash, and one for the 'Neopixel', Adafruit's
+name for a WS2812 RGB LED.
+
+Pins PA30-PA31 have pads on the underside of the board for the Serial Wire
+Debug Interface (SWCLK and SWDIO, respectively), allowing connection to a
+debug probe.
+
+The [pinouts](https://learn.adafruit.com/adafruit-feather-m4-express-atsamd51/pinouts)
+page in the User Guide has further information, photos, and a useful diagram
+showing the various pin functions. (This diagram does have some errors, however;
+the most noticeable being the labeling of several pins as outputs for TC6 & TC7,
+Timer/Counters which are not present on the chip.)
+
+### Board labeling
+
+Most pins broken out on the board are labeled Arduino-style, with A0-A5 having
+'analog' functions, and 0-13 (not all of which are available) for 'digital'.
+A few pins are labeled for specific functions, e.g. SPI, I2C, etc. All pins
+are aliased in 'board.hpp' to their GPIO equivalents; the 'digital' (0-13)
+pin aliases are prefixed with a 'D'.
+
+## Clocks
+
+Although the SAMD51 has pins available for crystal oscillators XOSC1 and XOSC2,
+there are no crystals installed on this board for their use. There is, however,
+a 32kHz crystal for the XOSC32K oscillator.
+
+Typically, an FDPLL is used as the system clock after bootup is complete; and
+using XOSC32K as a reference for it is not reliable. Even when using the workaround
+prescibed in Item 2.13 of the [Errata](https://ww1.microchip.com/downloads/aemDocuments/documents/MCU32/ProductDocuments/Errata/SAM-D5x-E5x-Family-Silicon-Errata-and-Data-Sheet-Clarification-DS80000748.pdf),
+the result is usually failure.
+
+The simplest solution is to source a GCLK generator with the 48MHz DFLL, and
+use a division factor that creates a suitable reference frequency for the FDPLL
+(e.g. 1-2MHz). This is the method used by Adafruit and TinyUSB in their
+FeatherM4 libraries, and that is what is done here.
+
+## Programming
+
+The Feather M4 has a UF2 bootloader. By using a utility to convert the '.elf'
+files (generated by scons or make) to '.uf2' files, it is then possible to 
+copy the uf2 file directly to the board.
+
+Adafruit has such a utility, called 'uf2conv.py', that works for this purpose
+(although the elf file needs to be converted to a binary or hex file first).
+The converter that is supplied with modm, 'elf2uf2.py', is ONLY for the rp2040,
+and will not work here.
```

Comment:
Why not? Is this a lack of configuration or are there different UF2 file formats?

Reply:
I did not spend too much time analyzing it. I had been using uf2conv.py for my own (non-modm) projects. Not being aware of the BOSSAC option, I tried the same strategy of creating a uf2 with modm; but at first it failed because the default 'float-abi' option is 'hard', and elf2uf2 kicked it out:

    if (e32h[3] & EF_ARM_ABI_FLOAT_HARD) != 0 :
        raise Exception("HARD-FLOAT not supported")

So I changed float-abi to 'soft', and it failed again (I don't recall the message). I found this near the start of elf2uf2:

uf2_config = {
    "rp2040": {
        "MAGIC_START0": 0x0A324655,
        "MAGIC_START1": 0x9E5D5157,
        "MAGIC_END": 0x0AB16F30,
        "FAMILY_ID": 0xe48bff56,
        "PAGE_SIZE": (1 << 8),
    }
}

And that's the only entry.

To me, that would seem to indicate that this was specifically written for the rp2040. There is a command line argument for 'target' and it appears that other chips could be added, but there's nothing there right now.

At that point, I rewrote the makefile (was using make at the time) to use uf2conv.py, and my problem was solved. I later changed to using BOSSAC, but my goal was to submit a board package to modm, and elf2uf2 was not helping.

The comments in my documentation were only intended to help people trying to use this board with modm, and if the uf2 converter was changed to handle more targets, they could easily be updated.

Reply:
Yes, good find! I didn't realize how much our uf2 converter was hardcoded. Let's fix it sometime later… the workaround with BOSSAc works well.